1.21.9 development

jpenilla opened the pull request:
Tracking issue: #12563

jpenilla committed:
Initial 1.21.9 update (squashed)

Co-authored-by: Bjarne Koll <git@lynxplay.dev>
Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>
Co-authored-by: Spottedleaf <Spottedleaf@users.noreply.github.com>

jpenilla committed:
Add Server#getRespawnWorld and Server#setRespawnWorld(World)

jpenilla committed:
move todos txt content into tracking issue

jpenilla committed:
Drop no-longer-needed packet diffs

We don't need to scale this anymore as vanilla has per-level world borders

jpenilla committed:
Use pitch when setting spawn position with Location

Owen1212055 committed:
Deprecate the Conversation API (for removal), Metadata API, and PlayerSpawnLocationEvent (for removal)

The Conversation API has been left to rot for the past few years and is considered dead. It does not support components and has had functionality ripped out over the years. Because of this, it is being deprecated for removal.

The metadata API is extremely misleading and causes memory leaks on every server I have seen use it. It just promotes bad practices and is misleading. The PDC system and just having their own solutions is better for developers.

Lastly deprecate PlayerSpawnLocationEvent for removal as events using this hack will be aggressively removed as we move towards better supporting the config API.

Spottedleaf committed:
Update Moonrise to 1.21.9-pre1

jpenilla committed:
Shift feature patch #s for Moonrise being first

jpenilla committed:
Update 'Optimize Network Manager and add advanced packet support'

jpenilla committed:
Update 'Allow Saving of Oversized Chunks'

jpenilla committed:
Update EAR 2.0 patch and enable oldPaperCommit for better 3-way merges

jpenilla committed:
Update 'Use Velocity compression and cipher natives'

jpenilla committed:
Update 'Optimize GoalSelector Goal.Flag Set operations'

jpenilla committed:
Update 'Optimize Voxel Shape Merging'

jpenilla committed:
Update 'Handle Oversized block entities in chunks'

jpenilla committed:
Update 'optimize dirt and snow spreading' and drop 'Optimise getChunkAt calls for loaded chunks' patch to code replaced by Moonrise

jpenilla committed:
Update 'Optimize Bit Operations by inlining', 'Remove streams from hot code'

jpenilla committed:
Update 'Optimize Pathfinder - Remove Streams / Optimized collections'

jpenilla committed:
Update 'Fix entity tracker desync when new players are added to the tracker'

jpenilla committed:
Update alternate redstone impls

jpenilla committed:
Update 'Improve exact choice recipe ingredients', 'Only write chunk data to disk if it serializes without throwing'

jpenilla committed:
Update 'Entity load/save limit per chunk', 'Attempt to recalculate regionfile header if it is corrupt'

jpenilla committed:
Update 'Incremental chunk and player saving', 'Optimise general POI access'

jpenilla committed:
Add some more Mannequin API

jpenilla committed:
Fix compile

Spottedleaf committed:
Replace ConcurrentUtil's Validate with Objects

Validate will be removed, and the methods in Objects are a
suitable replacement.

Spottedleaf committed:
Rewrite ticket counter system

The old method of tracking the counts was to copy a
Long2IntOpenHashMap each time a count was removed or added. This
was under the assumption that tickets with counter types were not
being added or removed frequently.

The new system uses ConcurrentLong2LongChainedHashTable instead
to support concurrently updating ticket counters for different
positions.
We copy the map when retrieving the forced chunks, however this
appears to only be used in the ForceLoad command so this shouldn't
be an issue.

Additionally, the ticket counts were not updated if the ticket
were to expire or if the ticket was UNKNOWN and added during
ChunkHolderManager#removeTicket(). These issues are now resolved.

Spottedleaf committed:
Drop "Optimize Collision to not load chunks"

Covered by Moonrise

Spottedleaf committed:
Apply "Flush regionfiles on save configuration option"

Spottedleaf committed:
Apply "Optimise collision checking in player move packet handling"

Spottedleaf committed:
Apply "Improve keepalive ping system"

Spottedleaf committed:
Apply "Optimise EntityScheduler ticking"

Spottedleaf committed:
Apply "Optional per player mob spawns"

Spottedleaf committed:
Apply "Improve cancelling PreCreatureSpawnEvent with per player mob spawns"

Spottedleaf committed:
Apply "Optimize Hoppers"

Lulu13022002 committed:
update generator

Couple of changes coming from the other branch/PRs:
- add support for defining registry constants out
of their value class (already required by at least 4 PRs).
- remove version hint in the generated annotation (to avoid noise during updates)
- basic support for parameterized types in API
- tweaks bootstrap to allow usage of dimension registries

Lulu13022002 committed:
add javadocs for new api

and fix some typos

Lulu13022002 committed:
Call BlockRedstoneEvent for the shelf

Lulu13022002 committed:
replace requireNonNull by checkArgument and deprecate MCUtil#toLocation

Lulu13022002 committed:
bump built-in datapack version

Lulu13022002 committed:
deprecate #lanterns material tag in favour of vanilla tag

jpenilla committed:
Mannequin#createDefaultProfile -> Mannequin#defaultProfile

ResolvableProfile is immutable

jpenilla committed:
Also rename Mannequin profile getter/setters for consistency

jpenilla committed:
Fix default respawn dimension being null for new worlds

Spottedleaf committed:
Use 0.0.4 release of ConcurrentUtil

https://github.com/Tuinity/Moonrise/commit/bab92354c01990ae54d734bc4395873918a330ca

lynxplay committed:
1.21.9-pre2

lynxplay committed:
1.21.9-pre2 feature patches

lynxplay committed:
Identifying comments for MinecraftServer

lynxplay committed:
Some diff improvements in prepare spawn task

Owen1212055 committed:
Move some stuff to ATs

lynxplay committed:
Drop no longer applicable ATs

Lulu13022002 committed:
bump api-version/datapack format and add a copper torch tag

Lulu13022002 committed:
generate ClientOption enums and mark CraftBlockData impl as null marked

jpenilla committed:
Add pre2 to api version

This is similar to what we did with 1.18-rc3 builds

jpenilla committed:
Set build channel to alpha

kennytv committed:
Port most of the anti-xray patch

jpenilla committed:
Disable oldPaperCommit to fix CI build

jpenilla committed:
Port the last Anti-XRay hunk

Spottedleaf committed:
Use ConcurrentUtil 0.0.5

For
https://github.com/Tuinity/ConcurrentUtil/commit/c3293d1d16b0196cbc8dc7ee8e813fbfbe9caa30

Lulu13022002 committed:
fix end portal teleportation

jpenilla committed:
[ci skip] Adjust spawn location event docs

Lulu13022002 committed:
simplify overloads

jpenilla committed:
Enable Anti-Xray

jpenilla committed:
Add AsyncPlayerSpawnLocationEvent#isNewPlayer (#13081)

jpenilla committed:
Fixup vanilla code for Anti-Xray changes

jpenilla committed:
Add back null check for chunkPacketBlockController

This is needed due to a circular reference during controller init

jpenilla committed:
Update to 1.21.9-pre3

jpenilla committed:
Add ResolvableProfile.SkinPatch (#13091)

jpenilla committed:
Update to 1.21.9-pre4

includes parchment update

Spottedleaf committed:
Update DataConverter to 1.21.9-pre4

https://github.com/PaperMC/DataConverter/commit/a3321c9b42d8a5f546ea99adcf047ad6e75a80a0
https://github.com/PaperMC/DataConverter/commit/9c1546f41101ab2b32f9d11a41d1a4742cd96df6

Spottedleaf committed:
Update Starlight stored version

https://github.com/Tuinity/Moonrise/commit/d2d297b804b98faeb5aab80c73aa247afe3806df

jpenilla committed:
Update publishing endpoint

jpenilla committed:
Call EntityChangeBlockEvent for copper golem statue changes (#13090)

jpenilla committed:
Update adventure

jpenilla committed:
Update to 1.21.9-rc1

jpenilla committed:
Fix empty SkinPatch initialization and swapped dynamic/static heuristic in ResolvableProfile build method

4drian3d committed:
Fixed modern forwarding support (#13098)

Lulu13022002 committed:
fix PlayerGameModeChangeEvent#cancelMessage

Lulu13022002 committed:
use PlayerShieldDisableEvent#getCooldown

jpenilla committed:
Fix a couple PlayerProfile mutability issues

Spottedleaf committed:
Remove parallel generation config option

https://github.com/Tuinity/Moonrise/commit/22e7dd06e87690e43619de1fe06adf349d5cca3f

Spottedleaf committed:
Use ThreadLocal for Starlight and ticket level propagator cache

https://github.com/Tuinity/Moonrise/commit/46d3cbf1eaa159140777a8addc5a44c874083b57

Spottedleaf committed:
Resize propagator queues less aggressively

https://github.com/Tuinity/Moonrise/commit/cca38110864f65ab3dd0d7ad1a421ee639fc14fb

Emilxyz committed:
Add GAMEMODE_SWITCHER gamemode change cause (#13097)

Co-authored-by: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>

Lulu13022002 committed:
remove old configs

Lulu13022002 committed:
remove api scheduled for removal

jpenilla committed:
Expose ResolvableProfile for Skull blocks (#13095)

Spottedleaf committed:
Update DataConverter to 1.21.9-rc1

https://github.com/PaperMC/DataConverter/commit/3138b268fdb26bccdb8d8c78e6549ba2807216ad

jpenilla committed:
Update Gradle wrapper to 9.1.0

jpenilla committed:
Update paperweight to 2.0.0-beta.19

granny committed:
Add spawnreason for building copper golems (#13112)

kennytv committed:
Update to 1.21.9

kennytv committed:
[ci skip] Update README

jpenilla committed:
Expose Copper Golem weathering state to API (#13089)

yannicklamprecht commented on paper-server/patches/sources/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java.patch:
```diff
@@ -1,5 +1,25 @@
+ public class CopperGolem extends AbstractGolem implements ContainerUser, Shearable {
+-    private static final long IGNORE_WEATHERING_TICK = -2L;
+-    private static final long UNSET_WEATHERING_TICK = -1L;
++    public static final long IGNORE_WEATHERING_TICK = -2L; // Paper - public
```
Move that to ATs

kennytv replied:
Doing in a patch for single fields/variables in low-conflict areas is easier than adding ATs to the separate file

yannicklamprecht commented on paper-server/patches/sources/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java.patch:
```diff
@@ -1,5 +1,25 @@
+     @Nullable
+     private UUID lastLightningBoltUUID;
+-    private long nextWeatheringTick = -1L;
++    public long nextWeatheringTick = -1L; // Paper - public
```
Here as well.

Lulu13022002 committed:
update enchantment target and rework some component impl

Lulu13022002 committed:
fix NPE in legacy profiles

Closes https://github.com/PaperMC/Paper/issues/13120

Lulu13022002 committed:
remove old test

Lulu13022002 committed:
fix copper chest inventory

Closes https://github.com/PaperMC/Paper/issues/13117

jpenilla committed:
Add ItemTransportingEntityValidateTargetEvent closes #13114 (#13115)

lynxplay committed:
Correctly reset quick crafting state on out of order packet

Correctly emit a inventory click event for out-of-order click packets to
prevent ghost items.

Lulu13022002 committed:
add shelf inventory type

Lulu13022002 committed:
validate mannequin pose

Commit:
Fix command-block-overrides from commands.yml
